File: app/src/main/java/com/iita/akilimo/adapters/FieldYieldAdapter.java
package com.iita.akilimo.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import com.iita.akilimo.R;
import com.iita.akilimo.entities.FieldYield;
import com.iita.akilimo.utils.ItemAnimation;
import com.iita.akilimo.utils.Tools;

import org.jetbrains.annotations.NotNull;

import java.util.List;

public class FieldYieldAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private List<FieldYield> items;

    private Context ctx;
    private OnItemClickListener mOnItemClickListener;
    private int animation_type;
    private int lastPosition = -1;
    private int rowIndex = -1;
    private boolean on_attach = true;
    private double selectedYieldAmount = 0.0;


    public interface OnItemClickListener {
        void onItemClick(View view, FieldYield fieldYield, int position);
    }

    public void setOnItemClickListener(final OnItemClickListener mItemClickListener) {
        this.mOnItemClickListener = mItemClickListener;
    }

    public FieldYieldAdapter(Context context, List<FieldYield> items, int animation_type) {
        this.items = items;
        ctx = context;
        this.animation_type = animation_type;
    }

    public void setItems(double selectedYieldAmount, @NonNull List<FieldYield> items) {
        this.items = items;
        this.selectedYieldAmount = selectedYieldAmount;
        notifyDataSetChanged();
    }

    public class OriginalViewHolder extends RecyclerView.ViewHolder {
        public ImageView rootYieldImage;
        public TextView name;
        public View layoutView;
        public CardView mainCard;

        public OriginalViewHolder(View v) {
            super(v);
            name = v.findViewById(R.id.name);
            rootYieldImage = v.findViewById(R.id.rootYieldImage);
            mainCard = v.findViewById(R.id.mainCard);
            layoutView = v.findViewById(R.id.lyt_parent);
        }
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder;
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_card_recommendation_image, parent, false);
        viewHolder = new OriginalViewHolder(view);
        return viewHolder;
    }

    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NotNull RecyclerView.ViewHolder holder, final int position) {
        if (holder instanceof OriginalViewHolder) {
            OriginalViewHolder view = (OriginalViewHolder) holder;

            FieldYield fieldYield = items.get(position);
            double currentYieldAmount = fieldYield.getYieldAmount();
            view.name.setText(fieldYield.getFieldYieldLabel());
            Tools.displayImageOriginal(ctx, view.rootYieldImage, fieldYield.getImageId());

            view.layoutView.setOnClickListener(view1 -> {
                if (mOnItemClickListener != null) {
                    mOnItemClickListener.onItemClick(view1, items.get(position), position);
                }
            });

            if ((rowIndex == position) || (currentYieldAmount == selectedYieldAmount)) {
                view.mainCard.setCardBackgroundColor(ctx.getResources().getColor(R.color.green_100));
            } else {
                view.mainCard.setCardBackgroundColor(ctx.getResources().getColor(R.color.grey_3));
            }

            setAnimation(view.itemView, position);
        }
    }

    public void setActiveRowIndex(int position) {
        rowIndex = position;
    }

    @Override
    public int getItemCount() {
        return items.size();
    }


    private void setAnimation(View view, int position) {
        if (position > lastPosition) {
            ItemAnimation.animate(view, on_attach ? position : -1, animation_type);
            lastPosition = position;
        }
    }

}

File: app/src/main/java/com/akilimo/mobile/adapters/FieldYieldAdapter.kt
package com.akilimo.mobile.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.akilimo.mobile.R
import com.akilimo.mobile.databinding.ItemCardRecommendationImageBinding
import com.akilimo.mobile.entities.FieldYield
import com.akilimo.mobile.utils.TheItemAnimation
import com.akilimo.mobile.utils.TheItemAnimation.animate
import com.akilimo.mobile.utils.Tools.displayImageOriginal

class FieldYieldAdapter(
    private val animationType: Int = TheItemAnimation.SCALE,
    private val showImage: Boolean = true,
    private val isItemSelected: (FieldYield) -> Boolean,
    private val onItemClick: (View, FieldYield, Int) -> Unit
) : ListAdapter<FieldYield, FieldYieldAdapter.ViewHolder>(FieldYieldDiffCallback()) {


    class ViewHolder(val binding: ItemCardRecommendationImageBinding) :
        RecyclerView.ViewHolder(binding.root)


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = ItemCardRecommendationImageBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val context = holder.itemView.context
        val fieldYield = getItem(position)

        with(holder.binding) {
            recImgTitle.text = fieldYield.fieldYieldLabel

            if (showImage) {
                displayImageOriginal(context, recImgImage, fieldYield.imageId)
                recImgImageContainer.visibility = View.VISIBLE
            } else {
                recImgImageContainer.visibility = View.GONE
            }

            val isSelected = isItemSelected(fieldYield)
            val colorRes = if (isSelected) R.color.green_100 else R.color.grey_3
            recImgCard.setCardBackgroundColor(ContextCompat.getColor(context, colorRes))

            recImgCard.setOnClickListener { view ->
                onItemClick(view, fieldYield, position)
            }

            animate(holder.itemView, position, animationType)
        }
    }

    private class FieldYieldDiffCallback : DiffUtil.ItemCallback<FieldYield>() {
        override fun areItemsTheSame(oldItem: FieldYield, newItem: FieldYield): Boolean {
            return oldItem.yieldAmount == newItem.yieldAmount
        }

        override fun areContentsTheSame(oldItem: FieldYield, newItem: FieldYield): Boolean {
            return oldItem == newItem
        }
    }
}
